index.rst

```diff
@@ -48,10 +48,109 @@
+
+Notebooks provide a nice environment for exploratory and pedagogical code development.
+Some reasons they are good for this type of work are:
+
+- They incorporate tab completion and easy access to help information
+- Documentation can reside with the code via markdown cells
+- Cells can be executed in arbitrary order, so the entire script doesn't need to be run to iterate on a particular part of the notebook
+- Code in cells can be flexible.  E.g. one cell can execute bash commands while the next cell contains python
+
+The same reasons notebooks perform well in exploratory scenarios are the reasons they are somewhat non-optimal for rigorous software engineering.
+A concrete example is that running cells in arbitrary order makes it easy to end up with a notebook in a state where it will not successfully complete when run sequentially.
+This makes testing tricky.
+
+Commissioning and science validation are, by their nature, exploratory, analytical exercises.
+This makes notebooks a natural place to start.
+However, we expect initial exploratory efforts to result in tooling we will want to execute over and over, including potentially in a batch system.
+With these two competing desiderata, we need a plan for how to transition between the two contexts: exploration vs. concrete tooling.
+
+A plan
+======
+
+I will not go into aspects that are part of the standard DM development workflow here.
+Instead I will focus on the aspects specific to development with notebooks.
+
+Repository organization
+-----------------------
+As a general rule, notebooks should be kept in a single git repository for each broad activity.
+E.g. a repository for commissioning and a repository for DM Science Validation.
+
+The repositories should have a defined structure.
+Informed by the data management git repositories, I suggest the following:
+
+- ``notebooks/wip`` -- A directory that holds notebooks that are works in progress.
+  These need not adhere to any specific quality specifications, but should have enough documentation to be readable by someone else.
+- ``notebooks/prod`` -- Notebooks meant to adhere to specific quality guidelines.
+  Notebooks cannot be added to ``prod`` from ``WIP`` without a review.
+- ``bin`` -- A place to put command line tools.
+    Command line tools are constructed from python modules as much as possible to improve test coverage.
+    I.e. any code logic should be factored into functions/methods in modules in the ``python`` directory.
+- ``python`` -- Python library code goes here.
+- ``tests`` -- A place for unit tests of the modular python code.
+  
+These repositories will be ``eups`` packages so that we can use all the existing infrastructure.
+
+Notes on notebook review
+------------------------
+
+Notebooks should be reviewed by someone other than the original author.
+The review should look for several things:
+
+- The notebook should adhere to a standard templated format (not defined here).
```

Comment:
Is one defined elsewhere that could be linked to? Stack-club have developed a nice template format, could we adopt it?

Reply:
I don't think it is templated, but I can look.